src/server/transaction.cc

```diff
@@ -399,10 +399,12 @@ OpStatus Transaction::InitByArgs(Namespace* ns, DbIndex index, CmdArgList args)
   }
 
   if ((cid_->opt_mask() & CO::NO_KEY_TRANSACTIONAL) > 0) {
-    if ((cid_->opt_mask() & CO::NO_KEY_TX_SPAN_ALL) > 0)
+    if (((cid_->opt_mask() & CO::NO_KEY_TX_SPAN_ALL) > 0) || IsActiveMulti()) {
```

Comment:
can you please add a comment for these conditions?

Reply:
I suspect it's sub-optimal. I suspect that we resize shard data when we schedule the lock ahead transactions. I agree that we should not shrink it. If we revert this condition the test you added will fail?

@dranikpg  Can you please review the change in this function specifically?

Reply:
Surely we can't enable all shards, it's a misunderstanding of what goes on

We use the NO_KEY_TRANSACTIONAL flag to make commands without keys transactional. Why? To replicate them via the journal. In fact, it was Adi who introduced this 🙂  https://github.com/dragonflydb/dragonfly/pull/1114

Putting order aside, any shard fits us to transfer replication data. So for active multi, let's find the first shard that has keylocks aquired and enable it. The only question now is linearizability - if we use differnet shards for script loads, they can arrive in a different order

If that's a problem, we have to always schedule on shard X if we find SCRIPT LOAD, or make it global.

--- 

By the way, it's a bug with NO_KEY_TX_SPAN_ALL as well... It's used by ft.search as a cheap read-only alternative to GLOBAL. I will take over the search issue.